### PR TITLE
Support v4 bundle export and import

### PR DESCRIPTION
**Pull request by hatched**
Machine placement directives changed in the v4 bundle spec. This implements that spec and makes sure that you can import the exported bundles.

### CODE REVIEW

**Comment by hatched** [comment]
#### To QA

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2497/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2498/
Test PASSed.

File: app/models/models.js [hunk 1]
```diff
@@ -2154,7 +2144,12 @@ YUI.add('juju-models', function(Y) {
...
-          machineName = 'lxc:' + machineNames[machine.parentId];
+          var machineId = machine.parentId;
+          if (machineNames[machineId]) {
+            machineName = 'lxc:' + machineNames[machineId];
+          } else {
+            machineName = 'lxc:' + machine.parentId;
+          }
         } else {
           // The "owner" is the service that deployer will use to allocate other
           // service units, e.g. "put this unit of mysql on the machine with
```
Comment (mitechie) [inline app/models/models.js]: this opens a quite but doesn't finish it? "put this unit....."

File: app/models/models.js [hunk 2]
```diff
@@ -2166,22 +2161,22 @@ YUI.add('juju-models', function(Y) {
...
           machineNames[machine.id] = machineName;
         }
 
         units.forEach(function(unit) {
           var serviceName = unit.service;
-          if (serviceName === owner) {
-            // Deployer doesn't allow placing units on a machine owned by a unit
-            // of the same service.
-            return;
+
```
Comment (mitechie) [inline app/models/models.js]: can you comment this a bit? I see what's up but I worry this will be more confusing in 6mo.

File: app/models/models.js [hunk 3]
```diff
@@ -2196,20 +2191,32 @@ YUI.add('juju-models', function(Y) {
...
-      var serviceList = this.services,
-          relationList = this.relations,
-          defaultSeries = this.environment.get('defaultSeries'),
-          result = {
-            envExport: {
-              services: {},
-              relations: []
-            }
-          };
+      var defaultSeries = this.environment.get('defaultSeries'),
```
Comment (mitechie) [inline app/models/models.js]: reminds me, can we make sure the default series of the fake backend is trusty now as a drive-by please?

File: app/models/models.js [hunk 4]
```diff
@@ -2286,15 +2293,39 @@ YUI.add('juju-models', function(Y) {
           serviceData.annotations = {'gui-x': anno['gui-x'],
```
Comment (mitechie) [inline app/models/models.js]: where did anno come from?

File: app/models/models.js [hunk 5]
```diff
@@ -2315,10 +2346,58 @@ YUI.add('juju-models', function(Y) {
         var relationData = endpoints.map(function(endpoint) {
           return endpoint[0] + ':' + endpoint[1].name;
         });
-        result.envExport.relations.push(relationData);
+        relations.push(relationData);
```
Comment (mitechie) [inline app/models/models.js]: there's another card about relations being exported when they're ghosts. Does this check and skip ghost services, machines, relations, and config?

File: app/models/models.js [hunk 6]
```diff
@@ -2315,10 +2346,58 @@ YUI.add('juju-models', function(Y) {
...
+      @return {String} The constraints in a string format.
+    */
+    _collapseMachineConstraints: function(constraints) {
+      var constraint = '';
+      var constraintMap = {
+        arch: 'arch',
+        cpuCores: 'cpu-cores',
+        cpuPower: 'cpu-power',
+        mem: 'mem',
+        disk: 'root-disk'
```
Comment (mitechie) [inline app/models/models.js]: is this defined else where? I worry this is a corner case where we dupe this list and have to maintain it.

File: app/utils/bundle-importer.js [hunk 7]
```diff
@@ -252,8 +260,17 @@ YUI.add('bundle-importer', function(Y) {
     */
     _execute_addMachines: function(record, next) {
       var parentKey = record.args[0].parentId;
+      parentKey = parentKey && parentKey.replace(/^\$/, '');
```
Comment (mitechie) [inline app/utils/bundle-importer.js]: what's a parent key that starts with a $ ?

File: test/test_model.js [hunk 8]
```diff
@@ -1924,7 +1927,8 @@ describe('test_model.js', function() {
       db.units.add(units, true);
       var placement = db._mapServicesToMachines(db.machines);
```
Comment (mitechie) [inline test/test_model.js]: this is a case of testing the private methods directly. It's good sometimes, but just pointing out as we talked about this today.

**Comment by mitechie** [comment]
Thanks for this change. It looks pretty good, however I'm a bit concerned with all the new lines of code without new tests?

For instance there's deduping code that doesn't have any tests that dupes are removed. There's the constraints code that doesn't have any tests against it that I can see here, and you enabled multiple units on an lxc now but no new tests verifying that works.

File: app/models/models.js (hunk 4 above)
Comment (hatched) [inline app/models/models.js]: 2 lines up :)

File: app/models/models.js (hunk 5 above)
Comment (hatched) [inline app/models/models.js]: This maintains the broken functionality. I was cutting scope of this branch. The card is still on the board.

File: app/models/models.js (hunk 6 above)
Comment (hatched) [inline app/models/models.js]: There is a list of constraints in the env go.js but not a map

File: app/utils/bundle-importer.js (hunk 7 above)
Comment (hatched) [inline app/utils/bundle-importer.js]: all 'key' references in the recordSet returned from bundlelib use that as an indicator that the field needs to be replaced by a value found in the specified keys record.

File: app/models/models.js (hunk 1 above)
Comment (hatched) [inline app/models/models.js]: ahh this is from the old comments I'll fix this

File: app/models/models.js (hunk 2 above)
Comment (hatched) [inline app/models/models.js]: Yeah I should do that

File: app/models/models.js (hunk 3 above)
Comment (hatched) [inline app/models/models.js]: It's currently precise. Changing this will not be a drive-by as it'll break a number of tests and change how the charmstore search results work. while in sandbox mode.

Comment (mitechie) [inline app/models/models.js]: ok added a card as a follow up. It's a bit past time we moved onto last year's release.

**Comment by makyo** [comment]
QA issue: adding machines without any constraints (i.e.: accepting your environment's default instance size) leads to those machines not showing up in the `machines` spec: https://gist.github.com/makyo/b0dd8d6637a6c18b4250 (Also, can we have the default file export name be `bundle.yaml` instead of `bundles.yaml` since we've gotten rid of the basket concept, now?)

**Comment by makyo** [comment]
NB placement looks okay otherwise, though we can probably simplify `lxc:wordpress/0` to just that machine's name (`lxc:1`).  They're equivalent, but one's a little more comprehensible.

**Comment by makyo** [comment]
Much better, my suspicion about the existing bug being the cause of the empty machines appears to have been correct: https://gist.github.com/makyo/6eceb81649c994dfb5d5  It's a little curious that machines where I didn't specify constraints during creation have them anyway, are those defaults in the GUI?  Trying to QA in a live env now

**Comment by makyo** [comment]
Actually, on closer inspection, I specified a root-disk size for the machine I specified constraints for, and that appears to have not made it into the constraints list.  No key errors in the console, though

**Comment by hatched** [comment]
I'm willing to chalk that up to an existing bug. The generation code simply takes what's in the constraints object and formats it out.

**Comment by makyo** [comment]
:+1:

**Comment by hatched** [comment]
Thanks for the reviews and qa! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/865

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/866

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/867

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge
